src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -236,6 +237,71 @@ public static string TranscodeHelper(ReadOnlySpan<byte> utf8Unescaped)
             }
         }
 
+        public static int TranscodeHelperWithCleanupOnFailure(ReadOnlySpan<byte> utf8Unescaped, Span<char> destination)
+        {
+            try
+            {
+#if BUILDING_INBOX_LIBRARY
+                return s_utf8Encoding.GetChars(utf8Unescaped, destination);
+#else
+                unsafe
+                {
+                    fixed (byte* srcPtr = utf8Unescaped)
+                    fixed (char* destPtr = destination)
+                    {
+                        return s_utf8Encoding.GetChars(srcPtr, utf8Unescaped.Length, destPtr, destination.Length);
+                    }
+                }
+#endif
+            }
+            catch (DecoderFallbackException dfe)
+            {
+                destination.Clear();
+
+                // We want to be consistent with the exception being thrown
+                // so the user only has to catch a single exception.
```

Comment:
Is this moved from elsewhere? If not, I think it's wrong.

https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=net-6.0

> The exception that is thrown when a method call is invalid for the object's current state.

or

> InvalidOperationException is used in cases when the failure to invoke a method is caused by reasons _**other than invalid arguments**_. Typically, it is thrown when the state of an object cannot support the method call.

(emphasis mine)

So InvalidOperationException is "the node type isn't a string-supporting type".  This is an exception about the content.  DecoderFallbackException is a very appropriate exception for that.  Or whatever the JSON reader throws for invalid JSON might also be valid.  But InvalidOperationException isn't (unless that was already its current behavior)

Reply:
It's been moved from elsewhere. I agree that throwing `InvalidOperationException` for this type of error is off, but we seem to be doing it consistently, e.g. when detecting invalid surrogate pairs:

https://github.com/dotnet/runtime/blob/810a7f9cd78d6611c0473940cb733231f3eabc06/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L386-L411

Probably a breaking change to alter the exception type at this point.